n_helpers.c

```diff
@@ -182,28 +182,33 @@ uint32_t NoteBinaryEncodedMaxLength(uint32_t len)
   @brief  Receive a large binary object from the Notecard's binary buffer
   @param  buffer A buffer to hold the binary object
   @param  bufLen The total length of the provided buffer
-  @param  dataLen An out parameter to hold the length of the decoded data in the
-                  output buffer.
+  @param  offset The offset to the unencoded binary data already residing on
+                 the Notecard
+  @param  dataLen [in/out] The length of the decoded data to fetch from the
+                  Notecard. If you wish to fetch the entire buffer from the
+                  given offset, set this value to `NOTE_C_BINARY_RX_ALL`. This
+                  parameter will return the bytes actually received from the
+                  Notecard, which is required when using `NOTE_C_BINARY_RX_ALL`.
   @returns  NULL on success, else an error string pointer.
-  @note  Buffers are decoded in place. The original contents of the buffer
-         will be modified.
+  @note  The buffer must be large enough to hold the encoded value of the
+         original contents at the requested offset and length.
   @note  To determine the necessary size for the buffer, use
-         NoteBinaryRequiredRxBuffer.
+         `NoteBinaryRequiredBuffer()`.
 */
 /**************************************************************************/
 const char * NoteBinaryReceive(uint8_t * buffer, size_t bufLen,
-                               size_t * dataLen)
+                               size_t offset, size_t * dataLen)
 {
-    size_t requiredRxBufLen = 0;
-    const char *err = NoteBinaryRequiredRxBuffer(&requiredRxBufLen);
-    if (err) {
-        return err;
+    // Validate parameter(s)
+    if (!buffer) {
+        NOTE_C_LOG_ERROR("NULL buffer");
+        return ERRSTR("NULL buffer", c_err);
     }
-    if (!requiredRxBufLen) {
-        NOTE_C_LOG_ERROR("no data on notecard");
-        return ERRSTR("no data on notecard", c_err);
+    if (!dataLen) {
+        NOTE_C_LOG_ERROR("dataLen cannot be NULL");
+        return ERRSTR("dataLen cannot be NULL", c_err);
     }
-    if (requiredRxBufLen > bufLen) {
+    if (bufLen < (NoteBinaryEncodedMaxLength(*dataLen) + sizeof('\n'))) {
```

Comment:
Why not use `NoteBinaryRequiredBuffer()` as described in the docs?

I'm using NoteBinaryRequiredBuffer from the `cobs` response but it's still telling me the buffer is too small. I changed `(NoteBinaryEncodedMaxLength(*dataLen) + sizeof('\n'))` to `NoteBinaryRequiredBuffer()` and it gets past that check, but later fails with

```
{"req":"card.binary.get","offset":0,"length":73,"crc":"0053:78FA2A08"}
{"err":"offset+length extends beyond what's in the binary I/O buffer"}
.pio/libdeps/debug/Blues Wireless Notecard/src/note-c/n_helpers.c:229 (ERROR) offset+length extends beyond what's in the binary I/O buffer
```
I assume then that the `cobs` response on `web.put` requests is the COBS-encoded size of the buffer? But then how do I determine the size of the actual data?

Reply:
It was out of my usual hesitation to call into the same API.

However, doing so helps to eliminate the bug (i.e. `\n` is interpreted as an `int` and `sizeof()` returns 4 instead of 1), so I changed my opinion.

Reply:
I did make this change, then in a subsequent commit I deleted the API it was relying upon, so I ended up fixing the bug and changing it back.